src/main/java/edu/cornell/kfs/fp/batch/service/impl/AccountingXmlDocumentDownloadAttachmentServiceImpl.java

```diff
@@ -0,0 +1,128 @@
+                    fileContents, attachmentType);
+            formFile.delete();
+            return attachment;
+        } else {
+            LOG.error("createAttachmentFromBackupLink, the form file is NULL");
+        }
+        
+        throw new IOException ("Unable to download attachment: " + accountingXmlDocumentBackupLink.getLinkUrl());
+    }
+
+    protected File downloadFile(AccountingXmlDocumentBackupLink accountingXmlDocumentBackupLink) {
+        Collection<WebServiceCredential> creds = getWebServiceCredtials(accountingXmlDocumentBackupLink);
+        File formFile = null;
+        try { 
+            Invocation request = buildClientRequest(accountingXmlDocumentBackupLink.getLinkUrl(), creds);
+            formFile = downloadFileFromWebResource(accountingXmlDocumentBackupLink.getFileName(), request);
+        } catch (URISyntaxException | IOException e) {
+            LOG.error("downloadFile, unable to download file.", e);
+        }
+        return formFile;
+    }
+
+    protected Collection<WebServiceCredential> getWebServiceCredtials(AccountingXmlDocumentBackupLink accountingXmlDocumentBackupLink) {
+        Collection<WebServiceCredential> webServiceCredentials = webServiceCredentialService
+                .getWebServiceCredentialsByGroupCode(accountingXmlDocumentBackupLink.getCredentialGroupCode());
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("getWebServiceCredtials, the group code is " + accountingXmlDocumentBackupLink.getCredentialGroupCode() + " for the file " + 
+                    accountingXmlDocumentBackupLink.getFileName());
+            if (CollectionUtils.isNotEmpty(webServiceCredentials)) {
+                webServiceCredentials.stream().forEach(cred -> LOG.debug("getWebServiceCredtials, found a credential key: " + cred.getCredentialKey()));
+            } else {
+                LOG.debug("getWebServiceCredtials, no credentials found");
+            }
+        }
+        return webServiceCredentials;
+    }
+
+    protected Invocation buildClientRequest(String url, Collection<WebServiceCredential> creds)
+            throws URISyntaxException {
+        URI uri = new URI(url);
+        Builder builder = getClient().target(uri).request();
+        if (CollectionUtils.isNotEmpty(creds)) {
+            for (WebServiceCredential cred : creds) {
+                builder.header(cred.getCredentialKey(), cred.getCredentialValue());
+            }
+        }
+        return builder.buildGet();
+    }
+
+    protected File downloadFileFromWebResource(String fileName, Invocation request) throws IOException {
+        File downloadfile = new File(fileName);
+        Response response = request.invoke();
+        if (response.getStatus() == Response.Status.OK.getStatusCode()) {
+            InputStream is = response.readEntity(InputStream.class);
+            byte[] byteArray = IOUtils.toByteArray(is);
+            FileOutputStream fos = new FileOutputStream(downloadfile);
```

Comment:
Do you need to write a temporary file for this step? If you can keep the contents as just a byte array, then you can pass a ByteArrayInputStream into the AttachmentService when you create the actual attachment. That way, the AttachmentService can use ClamAV to scan the attachment for viruses before it gets written to disk.

Reply:
good idea